File: api2/api_medidas.py
from fastapi import FastAPI, HTTPException, Depends, Header
from pydantic import BaseModel
from typing import Optional, List
import aiomysql
import jwt
import hashlib
import datetime

app = FastAPI()

# Configuración
SECRET_KEY = "TU_CLAVE_SECRETA_AQUI"
DB_CONFIG = {
    'host': 'localhost',
    'port': 3306,
    'user': 'usuario_base_de_datos',
    'password': 'contraseña_base_de_datos',
    'db': 'nombre_base_de_datos'
}

# Modelos Pydantic
class LoginRequest(BaseModel):
    cuenta: str
    clave_acceso: str

class Medida(BaseModel):
    id: int
    zona: int
    fecha: datetime.datetime
    tension: Optional[float]
    peso: Optional[float]
    estructura: Optional[str]
    observaciones: Optional[str]

# Función para conexión a base de datos
async def get_db():
    conn = await aiomysql.connect(**DB_CONFIG)
    try:
        yield conn
    finally:
        conn.close()

# Función para verificar token
def verify_token(token: str):
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
        return payload
    except jwt.ExpiredSignatureError:
        raise HTTPException(status_code=401, detail="Token expirado.")
    except jwt.InvalidTokenError:
        raise HTTPException(status_code=401, detail="Token inválido.")

# Endpoint de login
@app.post("/login")
async def login(request: LoginRequest, db=Depends(get_db)):
    async with db.cursor(aiomysql.DictCursor) as cursor:
        await cursor.execute(
            "SELECT * FROM USUARIOS WHERE cuenta = %s AND clave_acceso = %s",
            (request.cuenta, hashlib.md5(request.clave_acceso.encode()).hexdigest())
        )
        user = await cursor.fetchone()

    if not user:
        raise HTTPException(status_code=401, detail="Credenciales inválidas.")

    payload = {
        "sub": user['id'],
        "cuenta": user['cuenta'],
        "exp": datetime.datetime.utcnow() + datetime.timedelta(hours=1)
    }
    token = jwt.encode(payload, SECRET_KEY, algorithm="HS256")

    return {"token": token}

# Endpoint para consultar medidas
@app.get("/medidas", response_model=List[Medida])
async def get_medidas(
    fecha_inicio: str,
    fecha_fin: str,
    tension_min: Optional[float] = None,
    tension_max: Optional[float] = None,
    peso_min: Optional[float] = None,
    peso_max: Optional[float] = None,
    authorization: str = Header(None),
    db=Depends(get_db)
):
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="Token no proporcionado.")

    token = authorization.split(" ")[1]
    verify_token(token)

    query = """
        SELECT id, zona, fecha, tension, peso, estructura, observaciones
        FROM MEDIDAS
        WHERE fecha BETWEEN %s AND %s
    """
    params = [fecha_inicio, fecha_fin]

    if tension_min is not None:
        query += " AND tension >= %s"
        params.append(tension_min)
    if tension_max is not None:
        query += " AND tension <= %s"
        params.append(tension_max)
    if peso_min is not None:
        query += " AND peso >= %s"
        params.append(peso_min)
    if peso_max is not None:
        query += " AND peso <= %s"
        params.append(peso_max)

    query += " ORDER BY fecha ASC"

    async with db.cursor(aiomysql.DictCursor) as cursor:
        await cursor.execute(query, params)
        results = await cursor.fetchall()

    return results

File: api/ApiMedidas.php
<?php
require 'vendor/autoload.php';

use Firebase\JWT\JWT;
use Firebase\JWT\Key;
use Minishlink\WebPush\WebPush;
use Minishlink\WebPush\Subscription;

class ApiMedidas
{
    private $pdo;
    private $key = 'TU_CLAVE_SECRETA_AQUI'; // Cambia esto por una clave secreta segura

    public function __construct()
    {
        $host = 'localhost';
        $dbname = 'nombre_base_de_datos';
        $user = 'usuario_base_de_datos';
        $pass = 'contraseña_base_de_datos';

        $this->pdo = new PDO("mysql:host=$host;dbname=$dbname;charset=utf8mb4", $user, $pass);
        $this->pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    }

    public function login($cuenta, $clave_acceso)
    {
        $stmt = $this->pdo->prepare('SELECT * FROM USUARIOS WHERE cuenta = :cuenta AND clave_acceso = :clave_acceso');
        $stmt->execute([
            ':cuenta' => $cuenta,
            ':clave_acceso' => md5($clave_acceso)
        ]);

        $usuario = $stmt->fetch(PDO::FETCH_ASSOC);

        if (!$usuario) {
            http_response_code(401);
            echo json_encode(['error' => 'Autenticación fallida.']);
            exit;
        }

        $payload = [
            'iss' => 'tu-servidor',
            'aud' => 'tus-clientes',
            'iat' => time(),
            'exp' => time() + 3600,
            'sub' => $usuario['id'],
            'cuenta' => $usuario['cuenta']
        ];

        $jwt = JWT::encode($payload, $this->key, 'HS256');

        echo json_encode(['token' => $jwt]);
    }

    public function obtenerMedidas($token, $fecha_inicio, $fecha_fin, $tension_min = null, $tension_max = null, $peso_min = null, $peso_max = null)
    {
        try {
            $decoded = JWT::decode($token, new Key($this->key, 'HS256'));
            $id_usuario = $decoded->sub;
        } catch (Exception $e) {
            http_response_code(401);
            echo json_encode(['error' => 'Token inválido: ' . $e->getMessage()]);
            exit;
        }

        $query = "SELECT id, zona, fecha, tension, peso, estructura, observaciones
                  FROM MEDIDAS
                  WHERE fecha BETWEEN :fecha_inicio AND :fecha_fin";

        $params = [
            ':fecha_inicio' => $fecha_inicio,
            ':fecha_fin' => $fecha_fin
        ];

        if ($tension_min !== null) {
            $query .= " AND tension >= :tension_min";
            $params[':tension_min'] = $tension_min;
        }
        if ($tension_max !== null) {
            $query .= " AND tension <= :tension_max";
            $params[':tension_max'] = $tension_max;
        }
        if ($peso_min !== null) {
            $query .= " AND peso >= :peso_min";
            $params[':peso_min'] = $peso_min;
        }
        if ($peso_max !== null) {
            $query .= " AND peso <= :peso_max";
            $params[':peso_max'] = $peso_max;
        }

        $query .= " ORDER BY fecha ASC";

        $stmt = $this->pdo->prepare($query);
        $stmt->execute($params);

        $medidas = $stmt->fetchAll(PDO::FETCH_ASSOC);

        echo json_encode(['datos' => $medidas]);
    }
}
